### ISSUE TITLE
How to install the BIOS upgrade sp90168.exe from the HP web page?

### ISSUE THREAD

**Issue by eriksjolund** [eriksjolund/install-linux-hp-elitebook-850-g5#2]
Although it is possible to upgrade the BIOS version from within the BIOS (over the network), there seems to be a newer version of the BIOS available from the HP web page:

    HP Firmware Pack (Q78)  01.03.00 Rev.A       16.5 MB        22 aug 2018

https://support.hp.com/us-en/drivers/selfservice/hp-elitebook-850-g5-notebook-pc/18491276

The file provided

 https://ftp.hp.com/pub/softpaq/sp90001-90500/sp90168.exe

is meant to be used by Windows 10 (64-bit). 
Is it possible to extract the information with some Linux tool and install it?

**Comment by RussianNeuroMancer** [comment]
@eriksjolund extraction via 7z should work.

**Comment by eriksjolund** [comment]
Thanks for the tip. I tried this.

Installed the package **p7zip-full**

    root@laptop:~# apt-get install p7zip-full

Downloaded the exe file and then extracted

    linux@laptop:~$ wget https://ftp.hp.com/pub/softpaq/sp91501-92000/sp91951.exe
    linux@laptop:~$ mkdir /tmp/test
    linux@laptop:~$ cd /tmp/test
    linux@laptop:/tmp/test$ 7z e sp91951.exe

15 files got extracted.
I haven't tried anything more though.

**Comment by eriksjolund** [comment]
I haven't tried it on the EliteBook laptop yet but on
 an HP EliteDesk 800 G2 Tower PC, I tried the same approach (but with another downloaded sp*.exe file)

    linux@desktop:/tmp/test$ sudo mkdir -p /boot/efi/EFI/HP/BIOS/New
    linux@desktop:/tmp/test$ sudo ls /boot/efi/EFI/HP/BIOS/New | wc -l
    0
    linux@desktop:/tmp/test$ sudo cp *bin  /boot/efi/EFI/HP/BIOS/New
    linux@desktop:/tmp/test$ sudo ls /boot/efi/EFI/HP/BIOS/New | wc -l
    1
    linux@desktop:/tmp/test$

I rebooted and went into the BIOS and chose BIOS upgrade from local media. It worked.

**Comment by eriksjolund** [comment]
I documented the BIOS update on the HP EliteDesk 800 G2 Tower PC here:
https://askubuntu.com/a/1094838/20001

**Comment by eriksjolund** [comment]
I found some notes about the BIOS update for another HP EliteBook model:
https://lukas.dzunko.sk/index.php/Hardware:_How_to_update_BIOS_on_HP_Laptop_using_Linux
(In that example the extracted files are renamed)

**Comment by eriksjolund** [comment]
#### Short story: 
I gave it a new try but it failed

#### Long story: 
There is a new BIOS version (01.06.00 from 1 Feb 2019) that can be downloaded from the HP web page
so I gave it a new try.

```
linux@laptop:~$ mkdir /tmp/test
linux@laptop:~$ cd /tmp/test
linux@laptop:/tmp/test$ wget -q https://ftp.hp.com/pub/softpaq/sp94501-95000/sp94780.exe
linux@laptop:/tmp/test$ mkdir /tmp/extract
linux@laptop:/tmp/test$ cd /tmp/extract
linux@laptop:/tmp/extract$ 7z e /tmp/test/sp94780.exe 

7-Zip [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.UTF-8,Utf16=on,HugeFiles=on,64 bits,8 CPUs Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz (806EA),ASM,AES-NI)

Scanning the drive for archives:
1 file, 17359168 bytes (17 MiB)

Extracting archive: /tmp/test/sp94780.exe
--       
Path = /tmp/test/sp94780.exe
Type = PE
Physical Size = 17359168
CPU = x86
Characteristics = Executable 32-bit NoRelocs NoLineNums NoLocalSyms
Created = 2001-08-29 22:22:49
Headers Size = 4096
Checksum = 17423178
Name = stub32i.exe                                                 
Image Size = 290816
Section Alignment = 4096
File Alignment = 4096
Code Size = 73728
Initialized Data Size = 212992
Uninitialized Data Size = 0
Linker Version = 6.0
OS Version = 4.0
Image Version = 0.0
Subsystem Version = 4.0
Subsystem = Windows GUI
Stack Reserve = 1048576
Stack Commit = 4096
Heap Reserve = 1048576
Heap Commit = 4096
Image Base = 4194304
Comment = FileVersion: 4.0.100.1189
FileVersion: 01.06.00                                                    
ProductVersion: 4.0.0.0
ProductVersion: 01.06.00                                                    
Comments:                                                             
CompanyName: HPI                                                         
FileDescription:                                                             
InternalName: stub32                                                      
OriginalFilename: stub32i.exe                                                 
LegalCopyright:                                                             
ProductName: HP Firmware Pack (Q78)                                      
----
Path = [0]
Size = 17065896
Packed Size = 17065896
Virtual Size = 17065896
Offset = 278528
--
Path = [0]
Type = Cab
Offset = 23741
Physical Size = 17042153
Tail Size = 2
Method = MSZip
Blocks = 1
Volumes = 1
Volume Index = 0
ID = 12345

Everything is Ok         

Files: 15
Size:       38471904
Compressed: 17359168
linux@laptop:/tmp/extract$ md5sum *bin
a10ad77c7bc6d29aac7fa9e1374f84d1  Q78_010600.bin
linux@laptop:/tmp/extract$  
```

I copied the *.bin file into /boot/efi/EFI/HP/BIOS/New

```
root@laptop:~# cp /tmp/extract/*bin /boot/efi/EFI/HP/BIOS/New
root@laptop:~# 
```

After rebooting and going into the BIOS and then choosing 
BIOS upgrade from local media, it first seemed the BIOS file was detected. 
I was given the possibility to upgrade to 01.06.00  (if I remember correctly).
I chose that option, but within 1 second I got the error "_Invalid BIOS image_".

I also tried 
```
root@laptop:~# cp /tmp/extract/*inf /boot/efi/EFI/HP/BIOS/New/
```

(just a random guess because the files  Q78_010600.bin and Q78_010600.inf
had similar file names). That didn't help either.

**Comment by toddnni** [comment]
Thanks for the thread.

I needed to update the bios because of the sureview display and this bug <https://support.hp.com/za-en/document/c06137905> and I couldn't get the procedure you mentioned working either. 

However, over the network update in bios worked fine (Bios -> Update System Bios -> Check HP.com for BIOS Updates), and now the brightness setting in sureview mode works.

**Comment by doa379** [comment]
Did you manage by any chance to figure out why the BIOS reports invalid image when attempting to update from local media?

**Comment by eriksjolund** [comment]
No, I haven't figured out the why the BIOS reports invalid image when attempting to update from local media. 

But I have a guess, maybe it could be related to this:
When I installed Ubuntu on the laptop I erased everything on the SSD. As I understand now, the laptop came with an EFI partition with some HP software related to BIOS and firmware. Maybe that EFI partition is required for BIOS upgrades through local media?

**Comment by doa379** [comment]
> No, I haven't figured out the why the BIOS reports invalid image when attempting to update from local media.
> 
> But I have a guess, maybe it could be related to this:
> When I installed Ubuntu on the laptop I erased everything on the SSD. As I understand now, the laptop came with an EFI partition with some HP software related to BIOS and firmware. Maybe that EFI partition is required for BIOS upgrades through local media?

Yes I agree with that assessment. You need a EFI partition to do the BIOS upgrade.

If you're running any Linux distribution (including Ubuntu) then it's convenient to use an EFI partition for the boot loader instead of MBR.

If you can do this then be sure that the EFI partition is formatted vfat (fat16) and is at least 100MB in size. Get Hiron's Windows PE and copy the contents of the image to a bootable USB drive. Put the HP BIOS .exe on the drive too. Boot the PE image and flash your BIOS from that image by running the sp*.exe file. The update will write to the EFI boot partition in the HP/ directory with it's own BIOS update runtime program as well as actual BIOS .bin file. Follow the instructions and the HP BIOS update program will bootstrap its way at the next boot.

You can get the PE image from here:

https://www.hirensbootcd.org/download/

**Comment by songo** [comment]
Just for the record. Had "Invalid BIOS image" with unzipped  Q78_*.bin on HP 745 G6 also. The workaround provided by @doa379 worked.  After a closer look it became obvious that *.bin file written to EFI partition by Windows utility is different than one extracted manually using 7zip. They have different checksum and even different size. Hence the "Invalid…"  BIOS error. Unfortunately, I haven't found a way to obtain the correct firmware binary purely under Linux.

**Comment by jonlap** [comment]
Late to the party but anyway... Working procedure for Elitebook 850 G4, no EFI partition needed:
- Download bios update and extract the .bin file within
- Mount vfat formatted USB drive to /mnt, no matter if there's other files in it or not
- mkdir -p /mnt/Hewlett-Packard/BIOS/New
- Copy extracted .bin to /mnt/Hewlett-Packard/BIOS/New/
- In Elitebook bios, update from local media

If your USB drive is automatically mounted, adjust path (Hewlett-Packard directory needs to be in root of the USB drive). Hopefully this works with G5 also.

**Comment by doa379** [comment]
> Late to the party but anyway... Working procedure for Elitebook 850 G4, no EFI partition needed:
> 
>     * Download bios update and extract the .bin file within
> 
>     * Mount vfat formatted USB drive to /mnt, no matter if there's other files in it or not
> 
>     * mkdir -p /mnt/Hewlett-Packard/BIOS/New
> 
>     * Copy extracted .bin to /mnt/Hewlett-Packard/BIOS/New/
> 
>     * In Elitebook bios, update from local media
> 
> 
> If your USB drive is automatically mounted, adjust path (Hewlett-Packard directory needs to be in root of the USB drive).

Late to the party but never missed. That's not a G5 install procedure. The BIOS program was changed considerably from the jump between G4 and G5. Yes prior to G5 (G1-G4) you could do an upgrade off a USB drive.

**Comment by jonlap** [comment]
Ah, sorry for the noise then. I'll leave my previous comment as it is. This issue comes up in Google search when searching for G4 bios update from Linux so maybe it'll be helpful to someone.

About G5 firmware update (Q78_011500.bin from sp112373.exe) then. I compared Q78_011500.bin in the archive to firmware.bin that is created when recovery USB flash drive is made with HpFirmwareUpdRec64.exe. It appears that 640 bytes of something gets prepended to Q78_011500.bin by HpFirmwareUpdRec64.exe. That's the only difference between Q78_011500.bin and firmware.bin. What that 640 bytes is, I have no idea. Doesn't look like anything I'm able to recognize. I have that prepended part as a separate blob if someone want's to take a look (or update their firmware, ha). Not sure if it's OK to redistribute though, as it's not all zeroes. I'll probably take a closer look of that executable to see if I can find something in there.

**Comment by hamletmun** [comment]
As specified in the included "Bios Flash.htm", I put the renamed firmware as "HP\DEVFW\firmware.bin"
> These methods require that the firmware binary file is present on USB media or the hard drive in a specific folder - either "HP\DEVFW" or "EFI\HP\DEVFW" will support this feature. The file must be named 'firmware.bin'.

**Comment by sskras** [comment]
@hamletmun, and did it work?

**Comment by tukanos** [comment]
@sskras it did not work.  You either have to do it via windows or network update.

**Comment by hamletmun** [comment]
At least, it worked on HP EliteDesk 800 G4 SFF with 2.18, 2.19, 2.23, 2.24 and 2.25
(2.25 from https://support.hp.com/sk-en/document/ish_9818185-9818217-16/hpsbhf03891)

```
wget https://ftp.hp.com/pub/softpaq/sp149501-150000/sp149841.exe
7zz x sp149841.exe Q01_022500.bin
sudo mkdir -p /boot/efi/EFI/HP/DEVFW
sudo cp Q01_022500.bin /boot/efi/EFI/HP/DEVFW/firmware.bin
reboot
```

Press F10
Update System BIOS
Update System and Supported Device Firmware Using Local Media

**Comment by tukanos** [comment]
> At least, it worked on HP EliteDesk 800 G4 SFF with 2.18, 2.19, 2.23, 2.24 and 2.25 (2.25 from https://support.hp.com/sk-en/document/ish_9818185-9818217-16/hpsbhf03891)
> 
> ```
> wget https://ftp.hp.com/pub/softpaq/sp149501-150000/sp149841.exe
> 7zz x sp149841.exe Q01_022500.bin
> sudo mkdir -p /boot/efi/EFI/HP/DEVFW
> sudo cp Q01_022500.bin /boot/efi/EFI/HP/DEVFW/firmware.bin
> reboot
> ```
> 
> Press F10 Update System BIOS Update System and Supported Device Firmware Using Local Media

It would be great if it would work on my notebook.   For my HP Elitebook 850 G5 it did not work :(

**Comment by hamletmun** [comment]
USB stick with this path also worked
`HP\DEVFW\firmware.bin`

**Comment by tukanos** [comment]
> USB stick with this path also worked `HP\DEVFW\firmware.bin`

I have an issue trusting consistency of BIOS image of when copied to the flash memory. If the memory will go bad during the update, you will have expensive paperweight.

**Comment by hamletmun** [comment]
Both /boot/efi or /media/usb were using `vfat`
Did you check your filesystem?

**Comment by tukanos** [comment]
> Both /boot/efi or /media/usb were using `vfat` Did you check your filesystem?

I did:

NAME                     FSTYPE      FSVER    LABEL UUID                                   FSAVAIL FSUSE% MOUNTPOINTS
nvme0n1                                                                                                   
├─nvme0n1p1              vfat        FAT32          9DE1-20E0                               467.4M     9% /boot/efi

**Comment by hamletmun** [comment]
What is your current firmware version?
https://support.hp.com/us-en/drivers/swdetails/hp-elitebook-850-g5-notebook-pc/18491276/swItemId/ob-321768-1

Version: 01.11.01 Rev.A
- Fixes an issue where system firmware is updated from the recovery partition instead of the EFI partition.

**Comment by tukanos** [comment]
> What is your current firmware version? https://support.hp.com/us-en/drivers/swdetails/hp-elitebook-850-g5-notebook-pc/18491276/swItemId/ob-321768-1
> 
> Version: 01.11.01 Rev.A
> 
>     * Fixes an issue where system firmware is updated from the recovery partition instead of the EFI partition.

I have updated it already via network.  The upgrade process got interrupted so I was afraid I lost the mothermoard, but I managed to resuscitate it:

My bios version currently is: `Q78 Ver. 01.26.00`

**Comment by hamletmun** [comment]
With HP 800 G6, I can even use [fwupd](https://fwupd.org/lvfs/devices/com.hp.desktop.system.S01.firmware)
```
fwupdmgr get-upgrades
fwupdmgr upgrade
```

**Comment by tukanos** [comment]
> With HP 800 G6, I can even use [fwupd](https://fwupd.org/lvfs/devices/com.hp.desktop.system.S01.firmware)
> 
> ```
> fwupdmgr get-upgrades
> fwupdmgr upgrade
> ```

Well it is listed by fwupd as being supported.  Mine not.

**Comment by vc1cv1** [comment]
it's midnight, I want to go to bed.  I found my old usb stick that I had updated another HP model before.  Hopefully this saves someone some time in the future mimicking the folder structure.

This worked on a HP Elitebook x360 G2 (it was on bios P80_135 before).

so, whether you need the other HP*.exe files or not ...  I'm not going back to test. I initially did name it as 'firmware.bin' for the newer 151 on this usb stick -- that didn't work.  So, I just copied the file over as it was (P80_0151.bin) and it worked fine.

--------

./EFI
./EFI/HP
./EFI/HP/BIOS
./EFI/HP/BIOS/New
./EFI/HP/BIOS/New/P80_0135.bin
./EFI/HP/BIOS/New/HpqPswd.exe
./EFI/HP/BIOS/New/.BIN
./EFI/HP/BIOS/New/HPBIOSUPDREC.exe
./EFI/HP/BIOS/New/HPBIOSUPDREC64.exe
./EFI/HP/BIOS/New/HpqPswd64.exe
**./EFI/HP/BIOS/New/P80_0151.bin**
./EFI/HP/BIOS/New/firmware.bin
./Hewlett-Packard
./Hewlett-Packard/BIOS
./Hewlett-Packard/BIOS/New
./Hewlett-Packard/BIOS/New/P80_0135.bin
**./Hewlett-Packard/BIOS/New/P80_0151.bin**
./Hewlett-Packard/BIOS/New/firmware.bin
./boot
./boot/efi
./boot/efi/EFI
./boot/efi/EFI/HP
./boot/efi/EFI/HP/BIOS
./boot/efi/EFI/HP/BIOS/New
./boot/efi/EFI/HP/BIOS/New/.BIN
./boot/efi/EFI/HP/BIOS/New/HPBIOSUPDREC.exe
./boot/efi/EFI/HP/BIOS/New/HPBIOSUPDREC64.exe
./boot/efi/EFI/HP/BIOS/New/HpqPswd.exe
./boot/efi/EFI/HP/BIOS/New/HpqPswd64.exe
**./boot/efi/EFI/HP/BIOS/New/P80_0135.bin**
./boot/efi/EFI/HP/BIOS/New/P80_0135 (copy).bin
./boot/efi/EFI/HP/BIOS/New/firmware.bin
./boot/efi/EFI/HP/BIOS/New/P80_0151.bin

**Comment by ShoGinn** [comment]
Here is another option for those searching like me.. i could not get it to work with a USB stick so i just reverse engineered the online update:

https://gist.github.com/ShoGinn/e5a75126ec9780825cd39dbbce4e192d
